paparazzi/src/main/java/app/cash/paparazzi/internal/resources/PseudolocaleGenerator.kt

```diff
@@ -0,0 +1,144 @@
+import com.android.ide.common.resources.ResourceValueMap
+import com.android.ide.common.resources.configuration.FolderConfiguration
+import com.android.ide.common.resources.configuration.LocaleQualifier
+import com.android.resources.ResourceType
+import com.google.common.collect.Table
+
+private const val DO_NOT_TRANSLATE = "donottranslate"
+
+private val configCache = hashMapOf<Pair<ResourceSourceFile, LocaleQualifier>, RepositoryConfiguration>()
+
+fun String.pseudoLocalize(method: Pseudolocalizer.Method) =
+  if (isBlank()) {
+    this
+  } else {
+    with(Pseudolocalizer(method)) {
+      start() + text(this@pseudoLocalize) + end()
+    }
+  }
+
+fun Table<ResourceNamespace, ResourceType, ResourceValueMap>.pseudolocalizeIfNeeded(
+  localeQualifier: LocaleQualifier
+): Table<ResourceNamespace, ResourceType, ResourceValueMap> {
+  for (namespace in rowKeySet()) {
+    for (type in columnKeySet()) {
+      val resourceValues = this[namespace, type]!!
+      resourceValues.forEach { (key, value) ->
+        if (value !is BasicValueResourceItemBase || !value.isPseudolocalizable()) {
+          return@forEach
+        }
+
+        val method = when (localeQualifier.value) {
+          "en-rXA" -> Pseudolocalizer.Method.ACCENT
+          "ar-rXB" -> Pseudolocalizer.Method.BIDI
+          else -> return@forEach
+        }
+
+        var repositoryConfiguration = configCache[value.sourceFile to localeQualifier]
+        if (repositoryConfiguration == null) {
+          repositoryConfiguration = value.sourceFile.configuration.onNewLocaleQualifier(localeQualifier)
+          configCache[value.sourceFile to localeQualifier] = repositoryConfiguration
+        }
+
+        val pseudoLocaleSourceFile = value.sourceFile.onNewRepositoryConfiguration(repositoryConfiguration)
+        val pseudoItem: BasicValueResourceItemBase = when (value.resourceType) {
+          ResourceType.STRING -> pseudolocalizeString(
+            value as BasicValueResourceItem,
+            pseudoLocaleSourceFile,
+            method
+          )
+
+          ResourceType.PLURALS -> pseudolocalizePlural(
+            value as BasicPluralsResourceItem,
+            pseudoLocaleSourceFile,
+            method
+          )
+
+          ResourceType.ARRAY -> pseudolocalizeArray(
+            value as BasicArrayResourceItem,
+            pseudoLocaleSourceFile,
+            method
+          )
+
+          else -> return@forEach
+        }
+
+        resourceValues[key] = pseudoItem
+      }
+    }
+  }
+
+  return this
+}
+
+/**
+ * Check if the resource item is pseudolocalizable by verifying the source file name.
+ *
+ * Currently we cannot verify the translatable attribute from raw xml as it is not stored in BasicValueResourceItemBase.
+ * This will cause an issue that untranslatable strings are translated. The solution will be a new
+ * translatable Boolean field in BasicValueResourceItemBase, but this requires changes in Android Open Source Project.
```

Comment:
```suggestion
 * As a result, for now, untranslatable strings will be considered pseudolocalizable.
 * TODO: fix this
```

The AOSP approach is one way; we could also update the resource classes + repository loaders themselves to provide this feature.  There may be other ways, so better not to bake that into the source comments for now.